### PR TITLE
add `findall` implementation

### PR DESCRIPTION
**Pull request by zhenwu0728**
This PR port `findall` from [CUDA.jl](https://github.com/JuliaGPU/CUDA.jl/blob/master/src/indexing.jl).

### CODE REVIEW

File: src/indexing.jl [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+                b = indices[i]   # new position
+                ys[b] = i′
+            end
+
+            return
+        end
+
+        kernel = @metal name="findall" launch=false kernel(ys, bools, indices)
+        threads = Int(kernel.pipeline.maxTotalThreadsPerThreadgroup)
+        kernel(ys, bools, indices; threads)
```
Comment (maleadt) [inline src/indexing.jl:36]: You're not launching any groups?

File: src/indexing.jl [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+function Base.findall(bools::WrappedMtlArray{Bool})
+    I = keytype(bools)
+    indices = cumsum(reshape(bools, prod(size(bools))))
+
+    n = @allowscalar indices[end]
+    ys = MtlArray{I}(undef, n)
+
+    if n > 0
+        function kernel(ys::MtlDeviceArray, bools, indices)
+            i = (threadgroup_position_in_grid_3d().x - Int32(1)) * threads_per_threadgroup_3d().x + thread_position_in_threadgroup_3d().x
```
Comment (maleadt) [inline src/indexing.jl:23]: You're only using a 1D launch configuration, so no need for the 3D intrinsics.

File: test/array.jl [hunk 3]
```diff
@@ -404,4 +404,18 @@ end
...
+    @test testf(x->findall(x), rand(Bool, 100))
+    @test testf(x->findall(y->y>Float32(0.5), x), rand(Float32,100))
+
+    # ND
+    let x = rand(Bool, 10, 10)
+      @test findall(x) == Array(findall(MtlArray(x)))
+    end
+    let x = rand(Float32, 10, 10)
+      @test findall(y->y>Float32(0.5), x) == Array(findall(y->y>Float32(0.5), MtlArray(x)))
+    end
```
Comment (maleadt) [inline test/array.jl:418]: Better extend the problem sizes a little bit here so that it would have detected the issue of not launching any groups.

**Review by maleadt — CHANGES_REQUESTED** [review]

**Comment by zhenwu0728** [comment]
@maleadt Thanks for your quick response! I've made changes accrodingly.

**Review by maleadt — APPROVED** [review]

File: src/indexing.jl [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+using Base.Cartesian
+
+Base.to_index(::MtlArray, I::AbstractArray{Bool}) = findall(I)
+if VERSION >= v"1.11.0-DEV.1157"
+    Base.to_indices(A::MtlArray, I::Tuple{AbstractArray{Bool}}) = (Base.to_index(A, I[1]),)
+else
+    Base.to_indices(A::MtlArray, inds,
+                    I::Tuple{Union{Array{Bool,N}, BitArray{N}}}) where {N} =
+        (Base.to_index(A, I[1]),)
+end
```
Comment (maleadt) [inline src/indexing.jl:12]:
```suggestion

## logical indexing

# we cannot use Base.LogicalIndex, which does not support indexing but requires iteration.
# TODO: it should still be possible to use the same technique;
#       Base.LogicalIndex basically contains the same as our `findall` here does.

Base.to_index(::MtlArray, I::AbstractArray{Bool}) = findall(I)
if VERSION >= v"1.11.0-DEV.1157"
    Base.to_indices(A::MtlArray, I::Tuple{AbstractArray{Bool}}) = (Base.to_index(A, I[1]),)
else
    Base.to_indices(A::MtlArray, inds,
                    I::Tuple{Union{Array{Bool,N}, BitArray{N}}}) where {N} =
        (Base.to_index(A, I[1]),)
end
```

File: src/indexing.jl [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+function Base.findall(bools::WrappedMtlArray{Bool})
```
Comment (maleadt) [inline src/indexing.jl:14]:
```suggestion

## find*

function Base.findall(bools::WrappedMtlArray{Bool})
```

**Review by maleadt — APPROVED** [review]
As a minor it, I would just keep the comments from CUDA.jl. LGTM otherwise.

**Pull request merged**